service-worker.js

```diff
@@ -1,44 +1,94 @@
-async function sendUpdate(data) {
-  await fetch(
-    `http://localhost:9696/plugins/eeb9ac67-4aff-4b05-b6db-f5eb47d89974/${data.type}`,
-    {
-      method: "POST",
-      headers: {
-        "Content-Type": "application/json",
-      },
-      body: JSON.stringify(data.body),
+//https://stackoverflow.com/questions/51086688/mutex-in-javascript-does-this-look-like-a-correct-implementation
+class Mutex {
+    constructor() {
+        let current = Promise.resolve();
+        this.lock = () => {
+            let _resolve;
+            const p = new Promise((resolve) => {
+                _resolve = () => resolve();
+            });
+            // Caller gets a promise that resolves when the current outstanding
+            // lock resolves
+            const rv = current.then(() => _resolve);
+            // Don't allow the next request until the new promise is done
+            current = p;
+            // Return the new promise
+            return rv;
+        };
     }
-  );
 }
 
-chrome.runtime.onMessage.addListener(async function (
-  request,
-  sender,
-  sendResponse
-) {
-  await sendUpdate(request);
-});
+const baseUrl =
+    "http://localhost:9696/plugins/eeb9ac67-4aff-4b05-b6db-f5eb47d89974";
+
+const mutex = new Mutex();
+
+async function sendInitialData() {
+    let tabs = await chrome.tabs.query({});
+
+    await fetch(`${baseUrl}/setAllTabs`, {
+        method: "POST",
+        headers: {
+            "Content-Type": "application/json",
+        },
+        body: JSON.stringify(tabs),
+        signal: AbortSignal.timeout(200),
+    });
+}
+
+async function sendUpdate(endpoint, data) {
+    //Let's lock this to make sure the firstRequest feature works correctly.
+    //If we don't, weird stuff happens if we send 2 requests while waiting for the response and we lose data.
+    let unlock = await mutex.lock();
+
+    try {
+        let response = await fetch(`${baseUrl}/${endpoint}`, {
+            method: "POST",
+            headers: {
+                "Content-Type": "application/json",
+            },
+            body: JSON.stringify(data),
+            signal: AbortSignal.timeout(200),
+        });
+
+        let json = await response.json();
+
+        if (json.firstRequest) {
+            //if we get here, it means chrome
+            //was already open when artemis started.
+            //In this case, we need to send all the tabs,
+            //which will then be updated as they change.
+
+            await sendInitialData();
+        }
+    } catch (error) {
+        //TODO: do we even handle this? Ignoring is probably fine.
+    } finally {
+        unlock();
+    }
+}
 
-chrome.tabs.onUpdated.addListener(async function (tabId, changeInfo, tab) {
-  await sendUpdate({ type: "updatedTab", body: { tabId, changeInfo } });
+//TODO: @IceBot what is this for?
```

Comment:
This is for communications between the content script and Artemis. When the content script detects full screen, it will send a message to a service worker, and this service worker will send that data to Artemis.

Have you deleted the `raw` endpoint in the other PR? If so, this would stop working, and fullscreen would not be detected anymore.